### PR TITLE
[ENH, MRG] Handle Intracranial Electrode Coordinates from ``Raw`` in "head"

### LINKED ISSUE #856: Saving Intracranial Coordinates in BIDS
How should intracranial electrodes in "head" coordinates be handled? Should there be a `trans` argument in `write_raw_bids`? That would get to surface RAS but that doesn't look like one of the options in the error. Maybe that option needs to be added, seems like a lot of options though so someone may have thought about this thoroughly. To me, it makes sense to save them in surface RAS with a T1 in the BIDS directory, that seems about as easy to understand and reproducible as it can be. That would involve adding a `trans` argument and making whatever is the coordinate frame for surface RAS an acceptable option if a `trans` is provided.

Minimally reproducible script

```
import os
import os.path as op

import mne
import mne_bids

misc_path = mne.datasets.misc.data_path()
raw = mne.io.read_raw(op.join(misc_path, 'seeg', 'sample_seeg_ieeg.fif'))
tmp_dir = mne.utils._TempDir()
path = mne_bids.BIDSPath(root=tmp_dir, subject='sample', task='Motor')
mne_bids.write_raw_bids(raw, path)
```

Stack trace

```
Opening raw data file /home/alex/mne_data/MNE-misc-data/seeg/sample_seeg_ieeg.fif...
    Range : 1310640 ... 1370605 =   1311.411 ...  1371.411 secs
Ready.

Writing '/tmp/tmp_mne_tempdir_dhqc2o00/README'...

References
----------
Appelhoff, S., Sanderson, M., Brooks, T., Vliet, M., Quentin, R., Holdgraf, C., Chaumon, M., Mikulan, E., Tavabi, K., Höchenberger, R., Welke, D., Brunner, C., Rockhill, A., Larson, E., Gramfort, A. and Jas, M. (2019). MNE-BIDS: Organizing electrophysiological data into the BIDS format and facilitating their analysis. Journal of Open Source Software 4: (1896). https://doi.org/10.21105/joss.01896

Holdgraf, C., Appelhoff, S., Bickel, S., Bouchard, K., D'Ambrosio, S., David, O., … Hermes, D. (2019). iEEG-BIDS, extending the Brain Imaging Data Structure specification to human intracranial electrophysiology. Scientific Data, 6, 102. https://doi.org/10.1038/s41597-019-0105-7


Writing '/tmp/tmp_mne_tempdir_dhqc2o00/participants.tsv'...

participant_id	age	sex	hand
sub-sample	n/a	n/a	n/a

Writing '/tmp/tmp_mne_tempdir_dhqc2o00/participants.json'...

{
    "participant_id": {
        "Description": "Unique participant identifier"
    },
    "age": {
        "Description": "Age of the participant at time of testing",
        "Units": "years"
    },
    "sex": {
        "Description": "Biological sex of the participant",
        "Levels": {
            "F": "female",
            "M": "male"
        }
    },
    "hand": {
        "Description": "Handedness of the participant",
        "Levels": {
            "R": "right",
            "L": "left",
            "A": "ambidextrous"
        }
    }
}
Writing electrodes file to...  /tmp/tmp_mne_tempdir_dhqc2o00/sub-sample/ieeg/sub-sample_electrodes.tsv
Writing coordsytem file to...  /tmp/tmp_mne_tempdir_dhqc2o00/sub-sample/ieeg/sub-sample_coordsystem.json
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/alex/software/anaconda3/envs/mnedev/lib/python3.9/site-packages/mne_bids/write.py", line 1402, in write_raw_bids
    _write_dig_bids(bids_path, raw, overwrite, verbose)
  File "/home/alex/software/anaconda3/envs/mnedev/lib/python3.9/site-packages/mne_bids/dig.py", line 368, in _write_dig_bids
    coordsystem_path.update(space=coord_frame)
  File "/home/alex/software/anaconda3/envs/mnedev/lib/python3.9/site-packages/mne_bids/path.py", line 726, in update
    self._check()
  File "/home/alex/software/anaconda3/envs/mnedev/lib/python3.9/site-packages/mne_bids/path.py", line 842, in _check
    raise ValueError(f'space ({space}) is not valid for '
ValueError: space (CapTrak) is not valid for datatype (ieeg).
Should be one of ['ICBM452AirSpace', 'ICBM452Warp5Space', 'IXI549Space', 'fsaverage', 'fsaverageSym', 'fsLR', 'MNIColin27', 'MNI152Lin', 'MNI152NLin2009aSym', 'MNI152NLin2009bSym', 'MNI152NLin2009cSym', 'MNI152NLin2009aAsym', 'MNI152NLin2009bAsym', 'MNI152NLin2009cAsym', 'MNI152NLin6Sym', 'MNI152NLin6ASym', 'MNI305', 'NIHPD', 'OASIS30AntsOASISAnts', 'OASIS30Atropos', 'Talairach', 'UNCInfant', 'fsaverage3', 'fsaverage4', 'fsaverage5', 'fsaverage6', 'fsaveragesym', 'UNCInfant0V21', 'UNCInfant1V21', 'UNCInfant2V21', 'UNCInfant0V22', 'UNCInfant1V22', 'UNCInfant2V22', 'UNCInfant0V23', 'UNCInfant1V23', 'UNCInfant2V23', 'Other', 'Pixels', 'ACPC']
```

### PR DESCRIPTION
**Pull request by alexrockhill**
PR Description
--------------

Fixes #856.

I should have looked at what Adam had done, but it was quite complicated and lengthy dealing with `tsv`s. This isn't done but it might simplify things quite a bit to deal with the "head" space coordinates directly.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/main/doc/whats_new.rst) is updated
- [x] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

**Comment by alexrockhill** [comment]
cc @adam2392

**Comment by adam2392** [comment]
> I should have looked at what Adam had done, but it was quite complicated and lengthy dealing with `tsv`s. This isn't done but it might simplify things quite a bit to deal with the "head" space coordinates directly.

What are you referring to here?

I think the only qualm I could see the other devs having is the extra kwargs added. Is there a way to say attach the `trans` object to Raw without passing it in explicitly and same for the `mni_t` kwarg? So that way `write_raw_bids` only takes in the Raw object?

**Comment by alexrockhill** [comment]
> > I should have looked at what Adam had done, but it was quite complicated and lengthy dealing with `tsv`s. This isn't done but it might simplify things quite a bit to deal with the "head" space coordinates directly.
> 
> What are you referring to here?
> 
> I think the only qualm I could see the other devs having is the extra kwargs added. Is there a way to say attach the `trans` object to Raw without passing it in explicitly and same for the `mni_t` kwarg? So that way `write_raw_bids` only takes in the Raw object?

That's a good idea, I think you have to set the montage like you had done in the tutorial. The only thing is that in the `raw.info['chs'][idx]['coord_frame`]` will be "unknown" because MNE doesn't assign non-"head" coordinate frames to the `raw` object as far as I know.

I think the tutorial you had had better form assigning the `montage` i.e. I changed too much after thinking about it but I think the main improvement is dealing directly with `raw` and `montage` objects and not having to use `tsv`s but existing more within MNE objects.

**Comment by alexrockhill** [comment]
That's much better, I didn't actually change anything in `write.py`. Let me know what you think @adam2392. There is one hacky part where you have to remove the fiducials otherwise it will transform your "mni_tal" montage back to "head" when you `set_montage` but other than that looks super clean (and the current version only avoids this by not having fiducials in the first place).

**Comment by alexrockhill** [comment]
Oh also, `montage.apply_trans` is only on the development branch so that will require some dependency changes, cc @jasmainak

File: examples/convert_ieeg_to_bids.py [hunk 1]
```diff
@@ -70,65 +67,63 @@
...
 # %%
-# Now we make a montage stating that the iEEG contacts are in the MNI
-# coordinate system, which corresponds to the `fsaverage` subject in
-# FreeSurfer. For example, one can look at how MNE-Python deals with iEEG data
-# at `Working with SEEG
+# Now we make a montage in ACPC space. MNE stores channel locations
+# in the "head" coordinate frame, which has the origin at the center
+# between the left and right auricular points. BIDS requires that iEEG data
+# be in ACPC space so we need to convert from "head" to "mni_tal" which is
+# an `ACPC-aligned coordiante system
```
Comment (agramfort) [inline examples/convert_ieeg_to_bids.py:80]:
```suggestion
# an `ACPC-aligned coordinate system
```

File: examples/convert_ieeg_to_bids.py [hunk 2]
```diff
@@ -70,65 +67,63 @@
...
+# Now let's convert the montage to MNI Talairach ("mni_tal").
+montage = raw.get_montage()
+montage.apply_trans(mne.transforms.invert_transform(trans))  # head->mri
+montage.apply_trans(mri_mni_t)
+# a bit of a hack here; MNE will transform the coordinates to "head"
+# when you set the montage if there are fiducials and we don't want
+# that so we have to get rid of them
+montage.dig = montage.dig[3:]
+# warns that identity transformation to "head" is assumed which is what we want
+raw.set_montage(montage, verbose='error')
```
Comment (agramfort) [inline examples/convert_ieeg_to_bids.py:126]: we need to discuss this on discord with @larsoner

Comment (agramfort) [reply examples/convert_ieeg_to_bids.py:126]: this gymnastic is required by write_raw_bids for sEEG data?

Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:126]: Yes and no. Yes, the estimated fiducials and head->mri `trans` are more easily found this way for sEEG because there is no reason to have fiducials in your pipeline for sEEG so it makes sense to do them in an automated fashion. No, the transform to "head" is used properly, the only reason it is not in this instance (and wasn't before this PR, this PR just makes it explicit that we're suppressing the transformation to "head") is because we want to only pass a `raw` object with coordinate locations but those locations cannot be in "head" (unless we also supply a `trans` and an `mri_mni_t` (Talairach). I had it that way originally but it is very cumbersome and those arguments are unused for non-ieeg data.

File: examples/convert_ieeg_to_bids.py [hunk 3]
```diff
@@ -43,92 +42,99 @@
...
-montage = mne.channels.make_dig_montage(ch_pos=dict(zip(ch_names, elec)),
-                                        coord_frame='mni_tal')
-print(f'Created {len(ch_names)} channel positions')
-print(dict(zip(ch_names, elec)))
+# The ``fsaverage`` template brain is in MNI space so we can use the Talairach
+# transform to get to this space from "mri" or Freesurfer surface RAS space.
+# So, first we have to use the ``trans`` we found from assigning or estimating
+# fiducial points (see `Locating Intracranial Electrode Contacts
+# <https://mne.tools/dev/auto_tutorials/clinical/10_ieeg_localize.html>`_),
+# and then we can use the Freesurfer Talairach transform to get to MNI space.
```
Comment (adam2392) [inline examples/convert_ieeg_to_bids.py:91]:
```suggestion
# The ``fsaverage`` template brain is in MNI space so we can use the Talairach
# transform to get to this space from "mri" or Freesurfer surface RAS space.
# First we have to assign, or estimate the fiducial points (see `Locating Intracranial Electrode Contacts
# <https://mne.tools/dev/auto_tutorials/clinical/10_ieeg_localize.html>`_).
# Second, we create a Transformation object from ``head`` to ``mri`` coordinates using the fiducial points. 
# Finally, we can use the Freesurfer Talairach transform to get to MNI space.
```

File: examples/convert_ieeg_to_bids.py [hunk 4]
```diff
@@ -43,92 +42,99 @@
...
+
+# estimate trans
```
Comment (adam2392) [inline examples/convert_ieeg_to_bids.py:93]: "trans" seems like mne-python jargon. Maybe just "transformation from head to mri"?

File: examples/convert_ieeg_to_bids.py (hunk 2 above)
Comment (adam2392) [reply examples/convert_ieeg_to_bids.py:113]: Related issue I had a long time ago when things were just mapping from "unknown" to "head" within mne-python:

https://github.com/mne-tools/mne-python/issues/7691

**Comment by adam2392** [comment]
With the exception of the hack (https://github.com/mne-tools/mne-bids/pull/859#discussion_r689070104) on montages, this looks like a solid improvement to me.

I'm excited for MNE foraying into iEEG and overtaking Matlab there. Too bad just in time for me to finishing my phd :(

**Comment by alexrockhill** [comment]
> With the exception of the hack ([#859 (comment)](https://github.com/mne-tools/mne-bids/pull/859#discussion_r689070104)) on montages, this looks like a solid improvement to me.
> 
> I'm excited for MNE foraying into iEEG and overtaking Matlab there. Too bad just in time for me to finishing my phd :(

I saw that you saw the GUI, I think things are going well and it's almost everything you need for basic I/O. It would be nice to add a connectivity analysis example on the SEEG data that is localized if you want to collaborate on that :)

**Comment by alexrockhill** [comment]
Looks like the failure is unrelated, this is good to go by me

File: examples/convert_ieeg_to_bids.py (hunk 2 above)
Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:115]: Yeah I read the issue and I agree that ideally you wouldn't use "head" coordinates but like I said in another post, it's not the worst-defined coordinate frame because it's scan-agnostic (again in sEEG the lack of accuracy doesn't really make this worthwhile but I think in principle it's not too bad and it's not too bad to explain in tutorials to users). Especially if you can use the estimate from fsaverage fiducials, I think it's an okay solution.

**Comment by alexrockhill** [comment]
I'm a bit perplexed by the failed tests... but I just pushed again so maybe it was a weird one-time thing.

This is better and ready to go, I think. Besides that it would be nice to add how to save to BIDS in an individual subject's space. I guess if they used the Talairach you could get back exactly to the individual space but if they used a more accurate warp, the original coordinates would be lost. @agramfort just started more of a conversation about this so maybe that will come up with a clearer solution.

Well actually, I'll see about just adding an `acpc_aligned` flag to save as `ACPC` instead of `fsaverage` if the T1 is ACPC aligned.

**Comment by alexrockhill** [comment]
Ok, see how this looks, it won't build until https://github.com/mne-tools/mne-python/pull/9684 merges though...

File: examples/convert_ieeg_to_bids.py [hunk 5]
```diff
@@ -43,92 +41,63 @@
...
-# Let us confirm what our channel coordinates look like.
+# Now let's convert the montage to "mri"
+montage = raw.get_montage()
+montage.apply_trans(trans)  # head->mri
 
-# make a plot of the sensors in 2D plane
-raw.plot_sensors(ch_type='ecog')
+# careful, this should only be done in this specific circumstance,
+# it prevents MNE from transforming the coordinate system to "head"
+montage.remove_fiducials()
```
Comment (adam2392) [inline examples/convert_ieeg_to_bids.py:96]: I feel like the idea of having to remove fiducials just so one can obtain coordinates in "not-head" system seems weird to me? Wouldn't having more data points always be better?

I wasn't following other issues. Is it possible to make MNE not use "head" coordinate system?

File: examples/convert_ieeg_to_bids.py [hunk 6]
```diff
@@ -194,39 +164,23 @@
...
 
 # Now convert our data to be in a new BIDS dataset.
-bids_path = BIDSPath(subject=subject_id,
-                     task=task, acquisition="ecog", root=bids_root)
+bids_path = BIDSPath(subject=subject_id, task=task, root=bids_root)
 
 # write `raw` to BIDS and anonymize it into BrainVision format
 write_raw_bids(raw, bids_path, anonymize=dict(daysback=30000),
-               overwrite=True)
+               acpc_aligned=True, overwrite=True)
```
Comment (adam2392) [inline examples/convert_ieeg_to_bids.py:179]: If `acpc_aligned=True`, then do we need to `remove_fiducials` outside of `write_raw_bids`? Could we then do it inside the write function and just document this behavior?

File: examples/convert_ieeg_to_bids.py (hunk 5 above)
Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:96]: The way I understand it the fiducials are just what is used for coordinate transformation calculations so they're not like extra points in that regard. And then everything is standardized to head for things like plotting consistency but unless you want to pass the montage in addition, you're stuck with head. Passing the montage too might not be the worst solution.

File: examples/convert_ieeg_to_bids.py (hunk 6 above)
Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:179]: Might be better just to pass the montage as just mentioned in the previous comment but without that, no that wouldn't be possible because you have to remove before assigning to raw.

**Comment by alexrockhill** [comment]
Ok, I think that is a lot cleaner, maybe have a look when you have a second @adam2392.

Also, does anyone have any idea what the failed tests are from, I looked into it for a couple minutes and am very perplexed; there doesn't seem to be any problems with `mne.chpi` in 0.23...

**Comment by adam2392** [comment]
Perhaps some other ppl can look at this. I suppose I'm "okay" with the proposal given how complicated ieeg coordinates are.
Can you summarize why we need `acpc_aligned`? If you're passing in montage directly to `write_raw_bids`? I'm not following why that parameter is absolutely necessary but I might be overlooking something.

**Comment by alexrockhill** [comment]
> Perhaps some other ppl can look at this. I suppose I'm "okay" with the proposal given how complicated ieeg coordinates are.
> Can you summarize why we need `acpc_aligned`? If you're passing in montage directly to `write_raw_bids`? I'm not following why that parameter is absolutely necessary but I might be overlooking something.

You need `acpc_aligned` because you can't test whether their T1 is ACPC-aligned. If it's not the entire dataset is worthless...

**Comment by alexrockhill** [comment]
The montage is necessary because if you assign the montage to the `raw` with fiducials, MNE will transform it to head space. Since we don't want that, you have to pass the montage. Otherwise, we could remove the fiducials on the user-side in the tutorial but that is very hacky and should really be a backend thing

**Comment by alexrockhill** [comment]
Okay to merge? It's really not too drastic of a change, it just uses a `raw` object with coordinates instead of loading from a `tsv` and allows for the `ACPC` coordinate frame to be used which I think is ideal (storing in `fsaverage` is technically invertable *given the Talairach transform from the recon is used*, if SDR or another morphing method is used, it's not invertible so I would personally use subject-specific ACPC coordinates when possible).

File: examples/convert_ieeg_to_bids.py [hunk 7]
```diff
@@ -43,92 +41,58 @@
...
+# `ACPC-aligned coordinate system
+# <https://surfer.nmr.mgh.harvard.edu/fswiki/CoordinateSystems>`_.
+# We can either save the channel positions in the subject's anatomical
+# space (from their T1 image) or we can transform to a template space
+# such as ``fsaverage``. To save them in the individual space, it is
+# required that the T1 have been aligned to ACPC and then the channel positions
+# be in terms of that coordinate system. Automated alignment to ACPC has not
+# been implemented in MNE yet, so if the channel positions are not in
+# an ACPC-aligned coordinate system, using a template (like ``fsaverage``)
+# is the best option.
```
Comment (larsoner) [inline examples/convert_ieeg_to_bids.py:87]: In the email chain, it seems like people are advocating for using "other". If we can use "other" and set it to native/scanner RAS, then this is unambiguous and probably most widely understandable, as it's just related to `img.affine`, which is well understood and documented.

File: examples/convert_ieeg_to_bids.py [hunk 8]
```diff
@@ -144,12 +108,13 @@
...
 #
-# Currently, MNE-Python supports the ``mni_tal`` coordinate frame, which
-# corresponds to the ``fsaverage`` BIDS coordinate system. All other coordinate
-# frames in MNE-Python if written with :func:`mne_bids.write_raw_bids` are
-# written with coordinate system ``'Other'``. Note, then we suggest using
-# :func:`mne_bids.update_sidecar_json` to update the sidecar
-# ``*_coordsystem.json`` file to add additional information.
+# Currently, MNE-Python supports the ``mni_tal`` and ``mri`` coordinate frames,
+# corresponding to the ``fsaverage`` and ``ACPC`` BIDS coordinate systems
+# respectively. All other coordinate coordinate frames in MNE-Python if written
```
Comment (larsoner) [inline examples/convert_ieeg_to_bids.py:113]: I don't think that `MRI` in corresponds to `ACPC` BIDS coordinate frame in general. In general, `MRI` corresponds to FreeSurfer surface RAS / tkRAS. This will only be `ACPC` if the MRI has been processed so that FreeSurfer surface RAS is ACPC. So this dataset seems to be the exception to the rule, but this is phrased as if it's always the case...

Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:113]: I think that's explained pretty clearly immediately above

```
# %%
# When the locations of the channels in this dataset were found
# in `Locating Intracranial Electrode Contacts
# <https://mne.tools/dev/auto_tutorials/clinical/10_ieeg_localize.html>`_,
# the T1 was aligned to ACPC. So, this montage is in an
# `ACPC-aligned coordinate system
# <https://surfer.nmr.mgh.harvard.edu/fswiki/CoordinateSystems>`_.
# We can either save the channel positions in the subject's anatomical
# space (from their T1 image) or we can transform to a template space
# such as ``fsaverage``. To save them in the individual space, it is
# required that the T1 have been aligned to ACPC and then the channel positions
# be in terms of that coordinate system. Automated alignment to ACPC has not
# been implemented in MNE yet, so if the channel positions are not in
# an ACPC-aligned coordinate system, using a template (like ``fsaverage``)
# is the best option.
```

File: examples/convert_ieeg_to_bids.py (hunk 7 above)
Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:87]: -1 I don't see the merit in using coordinate system other in a tutorial about BIDS. I would rather just stick with `fsaverage` which at least is well-defined

Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:87]: When BIDS changes to allow `scanner RAS`, I'm happy to update the tutorial but until then, I think it should not be recommended in a tutorial

File: examples/convert_ieeg_to_bids.py (hunk 8 above)
Comment (larsoner) [reply examples/convert_ieeg_to_bids.py:113]: Agreed but reading *this section* the statement sounds global. I think it needs "for this dataset" or "for a ACPC-aligned dataset" or something

Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:113]: Sure just changed

**Review by larsoner — APPROVED** [review]
+1 for merge from me, but I don't understand all implications so others (@agramfort @jasmainak ) should review!

File: examples/convert_ieeg_to_bids.py (hunk 7 above)
Comment (jasmainak) [reply examples/convert_ieeg_to_bids.py:87]: I think we should go back to BIDS with the proposal if it's not accepted at the moment. There are lots of holes in the standard and that's why tool developers need to complete the development loop. In the past, I and @sappelhoff had to iterate quite a bit between developing MNE-BIDS, realizing that there is an issue then fixing it in the standard + validator, then fixing the datasets and finally incorporating that back in MNE-BIDS. I am worried that we'll develop unnecessarily complicated pipelines involving too many coordinate transforms otherwise (you may remember how Tal complained when we presented the anonymization procedure to her which was overly complex)

Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:87]: I hear you and I agree but I'm not sure how to get traction on the issue. It seems there is the MNE camp who thinks this is an issue and the everyone else camp who is happy to use "Other" and don't seem to be persuaded by the arguments in the discussion of this PR... not sure where to go from there. We've had @adam2392's issue up for a while, it has gone stale and we started an email thread with @agramfort which has also gone stale. I think the argument for scanner RAS is compelling but it hasn't been received that way so I'm not sure how to proceed in a different way to make that change

Comment (adam2392) [reply examples/convert_ieeg_to_bids.py:87]: @sappelhoff 's PR at https://github.com/bids-standard/bids-specification/pull/820 seems like it may address the concerns for having a "REQUIRED" pointer to the T1 image, if we choose to move forward with `scanner RAS` in BIDS, which encompasses the FreeSurfer T1.mgz, and possibly other reconstruction pipelines(?).

**Comment by adam2392** [comment]
I have updated the original issue in BIDS: https://github.com/bids-standard/bids-specification/issues/747#issuecomment-908580900

Please take a look and lmk if you see any issues / chime in.

Although I would like this PR merged for my sake, I do agree that if we can get a consensus in BIDS, then it'll make our lives that much easier down the line.

**Comment by jasmainak** [comment]
you're the best @adam2392 

let's move on here, but try to move the discussion on the BIDS front and circle back here to simplify later

File: examples/convert_ieeg_to_bids.py (hunk 7 above)
Comment (agramfort) [reply examples/convert_ieeg_to_bids.py:87]: honestly @alexrockhill I think you're doing a crazy gymnastic here but I don't think we should recommend. The data you use here is not ACPC aligned AFAIK so what @dorahermes says is that she would use "Other" and then refer to the T1.mgz origin coordinate system. That way simpler and not ambiguous (I think). Other is technically allowed by the standard and I prefer to add complexity to the code later than add it now and than possibly simplify later.

my 2c

Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:87]: Not sure what you mean, this data is ACPC-aligned. I aligned the T1 before finding the coordinates.

Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:87]: I think the only gymnastics are with the transformation to `fsaverage` with removing the fiducials. I agree that is not optimal but it seems like the only stable solution and *this PR isn't changing that, there were just no fiducials in the previous version so it wasn't explicit.*

Comment (agramfort) [reply examples/convert_ieeg_to_bids.py:87]: ok I had not realised these data were already ACPC aligned before running FS recon.

do you have time to discuss on discord?

Comment (jasmainak) [reply examples/convert_ieeg_to_bids.py:87]: Part of my objection to how iEEG is handled in MNE is that fiducials need to be defined, which does not make a lot of sense when there is no need for a "head"-device transform and causes unnecessary confusion. See also [comment by @adam2392](https://github.com/mne-tools/mne-bids/pull/859#discussion_r693286578) here. I didn't quite follow the argument why [montage needed to be in "head" coordinates because BIDS required ACPC](https://github.com/mne-tools/mne-bids/pull/859#issuecomment-905824219).

Isn't it possible to store electrode positions in `scanras` if iEEG (or a different coordinate system if explicitly defined by user), require no fiducials and internally in MNE handle the visualization based on what coordinate system montage claims to be in? See [here](https://gist.github.com/jasmainak/a6ce60f1e957fe27a5ead32d7efb81e4) some code for my  project ... there is no need to define fiducials manually. Robert seems to have come around to scanras by the way.

Comment (larsoner) [reply examples/convert_ieeg_to_bids.py:87]: > Isn't it possible to store electrode positions in scanras if iEEG (or a different coordinate system if explicitly defined by user), require no fiducials and internally in MNE handle the visualization based on what coordinate system montage claims to be in

Up until the last few months, no, this was not possible with FIF. Now it is possible but many MNE functions would need to be changed to accommodate this. This discussion has come up many times now. I always make the same argument (and have already here): to me it's at worst a minor inconvenience to have to define this coordinate frame (especially with `estimate_mni_fiducials` or whatever it's called) and have to pass around `trans`, and it's the path of least resistance to make things work properly in MNE until we can refactor a bunch of viz and forward functions to handle EEG-like electrode positions that are not always in a properly defined head coordinate frame.

Comment (larsoner) [reply examples/convert_ieeg_to_bids.py:87]: > I didn't quite follow the argument why montage needed to be in "head" coordinates because BIDS required ACPC.

In that quoted issue, there were two separate things: what a "head" coordinate system is, and what an ACPC coordinate system is. I think they are separate issues, I don't think a montage needs to be in head because BIDS required ACPC.

Comment (larsoner) [reply examples/convert_ieeg_to_bids.py:87]: ... it looks like @alexrockhill's previous answer to this question was in this PR in July for example:

https://github.com/mne-tools/mne-python/pull/9601/files#r675062176

> Yeah I agree it's not ideal but I think to be backward compatible with MEG/EEG, FIF requires it to be in the "head" frame. I've thought it a bit like how Freesurfer uses surface RAS as its internal representation frame, MNE uses "head". They are all arbitrary reference frames and at least there is a standard and you know how it's stored. Ideally the coordinate frame could be specified in the FIF but that seems like pretty large change.

@jasmainak let's not continue to discuss this here since I think of this as a "settled" issue (at least among @agramfort @adam2392 @alexrockhill ) even though it's not ideal, and mostly tangential to this PR. Feel free to open an MNE-Python issue where we can discuss further if you want, but please do search through previous discussions. Maybe it would help to have a Discord chat during office hours or something.

Comment (jasmainak) [reply examples/convert_ieeg_to_bids.py:87]: sounds good. I understand that it requires a major refactor in MNE but seems like a necessary step to properly integrate sEEG + not bog down the user with coordinate systems. The issue of coordinate transforms I already mention in the [sEEG issue](https://github.com/mne-tools/mne-python/issues/8751#issue-786490766) that was raised some time ago. 

> it's too verbose about transformations. I think it's just an identity transformation usually.

I was clearly confused what was going on in the example. Do you want me to start another issue or we can keep that one open, perhaps converting the bullet points into check boxes? I think in general starting from the example and making sure it's simple enough is a good way to go.

Comment (alexrockhill) [reply examples/convert_ieeg_to_bids.py:87]: I can open an issue and I'd be happy to work on it when I have time, I think it might be already in the FIF specification so we might just have to change MNE

**Review by agramfort — APPROVED** [review]
LGTM !

@adam2392 feel free to press the green button.

I think it's already a big step forward in the right direction.

thx @alexrockhill

**Comment by alexrockhill** [comment]
It might be nice actually to plot the T1 and show what ACPC-aligned looks like actually, let me add that real quick

**Comment by agramfort** [comment]
+1

>

**Review by adam2392 — APPROVED** [review]

**Comment by alexrockhill** [comment]
Artifact looks good to me https://5126-89170358-gh.circle-artifacts.com/0/dev/auto_examples/convert_ieeg_to_bids.html

**Comment by adam2392** [comment]
Thanks @alexrockhill !

**Comment by adam2392** [comment]
Wanna coordinate to take a stab at altering BIDS spec? :p

**Comment by sappelhoff** [comment]
Thanks for your work on this everybody. I am slowly catching up, the emerging solution looks good to me!
